src/PureSMT.hs

```diff
@@ -58,48 +58,33 @@ instance Default Options where
         numWorkers = Nothing
       }
 
-solve :: forall domain res. Solve domain => Ctx domain -> Problem domain res -> res
-solve = solveOpts @domain @res def
-
 -- | Evaluates a 'Problem' with an SMT solver chosen from a global pool of
 -- SMT solvers.
---
--- At least one pool is started for each value of @Ctx domain@. Note that
--- multiple pools of SMT solvers might be created for the same value of
--- @Ctx domain@ if @solve@ is called from different modules, or if GHC
--- optimizations fail to apply CSE over calls to solve in the same module.
-{-# NOINLINE solveOpts #-}
-solveOpts :: forall domain res. Options -> Solve domain => Ctx domain -> Problem domain res -> res
-solveOpts opts ctx = unsafePerformIO $ do
-  -- we end up with a list of MVars, which we will protect in another MVar.
-  allProcs <- initAll @domain opts ctx >>= newMStack
```

Comment:
Since we're propagating `IO` back into Pirouette's core, there's no need to call `initAll` when calling `solveOpts`, so I split these two functions.